# Jacob Evers

# Chapter 3 Data Preparation: Questions 21 - 25

In [97]:
# Import the packages that will be used to read and analyze the data
import pandas as pd
import numpy as np
from scipy import stats
# Import the data set
nutrition_subset = pd.read_csv(r"C:\Users\jacob\Downloads\nutrition_subset.txt")

<ol start = "21"> <em style="color:#3358FF";>
    <li> The elements in the data set are food items of various sizes, ranging from a teaspoon of cinnamon to an entire carrot cake.
        
<ol start = a> <li> Sort the data set by the saturated fat (saturated_fat) and produce a listing of the five food items highest in saturated fat.
    <li> Comment on the validity of comparing food items of different sizes.</em></ul></ol>

In [98]:
# Create a new list, sorted by in descending order by the saturated_fat column
nutrition_subset_sort = nutrition_subset.sort_values(['saturated_fat'], ascending = False)

In [99]:
# Display the top five items from the sorted list
nutrition_subset_sort.head(n=5)

food item  weight_in_grams  saturated_fat  \
378  CHEESECAKE                    1 CAKE             1110.0          119.9   
535  ICE CREAM; VANLLA; RICH 16% FT1/2 GAL            1188.0          118.3   
458  YELLOWCAKE W/ CHOCFRSTNG;COMML1 CAKE             1108.0           92.0   
581  CREME PIE                     1 PIE               910.0           90.1   
890  LARD                          1 CUP               205.0           80.4   

     cholesterol  
378         2053  
535          703  
458          609  
581           46  
890          195

<p style="color:#2CC531";> <b> Answer to 21.a </b> Comparing items of different sizes is not very useful. A person is not likely to consume an entire cheesecake or a full cup of lard in a single sitting. Instead, a more useful comparison would be to normalize the foods based on a normal serving size. Once normalized, then a comparison of the nutritional values would be much more informative.</p> 

<ol start="22"><em style="color:#3358FF";>
    <li>Derive a new variable, saturated_fat_per_gram, by dividing the amount of saturated fat by the weight in grams.
<ol start = a><li>Sort the data set by saturated_fat_per_gram and produce a listing of the five food items highest in saturated fat per gram.
    <li>Which food has the most saturated fat per gram? - <b style="color:#F0f71c";>Butter!</b>
        </em>
        </ol>
        </ol>

In [100]:
#Create a new column in the table called saturated_fat_per_gram
nutrition_subset['saturated_fat_per_gram'] =(nutrition_subset['saturated_fat'] / nutrition_subset['weight_in_grams']) 
# Sort the table in descending order by the saturated_fat_per_gram value
nutrition_subset_sort = nutrition_subset.sort_values(['saturated_fat_per_gram'], ascending = False)
# Display the first fifteen rows of the sorted table
nutrition_subset_sort.head(n=15)

food item  weight_in_grams  saturated_fat  \
908  BUTTER; SALTED                1 TBSP              14.00            7.1   
909  BUTTER; UNSALTED              1 TBSP              14.00            7.1   
710  BUTTER; UNSALTED              1/2 CUP            113.00           57.1   
709  BUTTER; SALTED                1/2 CUP            113.00           57.1   
913  BUTTER; UNSALTED              1 PAT                5.00            2.5   
912  BUTTER; SALTED                1 PAT                5.00            2.5   
899  LARD                          1 TBSP              13.00            5.1   
890  LARD                          1 CUP              205.00           80.4   
920  IMITATION CREAMERS; POWDERED  1 TSP                2.00            0.7   
210  CHOCOLATE; BITTER OT BAKING   1 OZ                28.35            9.0   
492  COCONUT; DRIED; SWEETND;SHREDD1 CUP               93.00           29.3   
576  COCONUT; RAW; PIECE           1 PIECE             45.00           13.4   
448  COCONUT; RAW; SHREDDED        1 CUP               80.00           23.8   
898  FATS; COOKING/VEGETBL SHORTENG1 TBSP              13.00            3.3   
907  FATS; COOKING/VEGETBL SHORTENG1 CUP              205.00           51.3   

     cholesterol  saturated_fat_per_gram  
908           31                0.507143  
909           31                0.507143  
710          247                0.505310  
709          247                0.505310  
913           11                0.500000  
912           11                0.500000  
899           12                0.392308  
890          195                0.392195  
920            0                0.350000  
210            0                0.317460  
492            0                0.315054  
576            0                0.297778  
448            0                0.297500  
898            0                0.253846  
907            0                0.250244

<ol start=23><em style="color:#3358FF";>
<li>Derive a new variable, cholesterol_per_gram.
    <ol start = a>
<li>Sort the data set by cholesterol_per_gram and produce a listing of the five food items highest in cholesterol fat per gram.
        <li>Which food has the most cholesterol fat per gram? - <b>Raw egg yolk</b></em>
    </ol>
    </ol>

In [101]:
# Create a cholesterol_per_gram value the same way we did with saturated_fat_per_gram
nutrition_subset['cholesterol_per_gram'] =(nutrition_subset['cholesterol'] / nutrition_subset['weight_in_grams']) 

# Sort the table in descending order by this new value
nutrition_subset_sort = nutrition_subset.sort_values(['cholesterol_per_gram'], ascending = False)

# Display the first fifteen rows of the newly sorted table
nutrition_subset_sort.head(n=15)
print(nutrition_subset_sort[])

food item  weight_in_grams  saturated_fat  \
119  EGGS; RAW; YOLK               1 YOLK               17.0            1.6   
58   CHICKEN LIVER; COOKED         1 LIVER              20.0            0.4   
45   BEEF LIVER; FRIED             3 OZ                 85.0            2.5   
167  EGGS; COOKED; FRIED           1 EGG                46.0            1.9   
186  EGGS; COOKED; HARD-COOKED     1 EGG                50.0            1.6   
184  EGGS; RAW; WHOLE              1 EGG                50.0            1.6   
185  EGGS; COOKED; POACHED         1 EGG                50.0            1.5   
189  EGGS; COOKED; SCRAMBLED/OMELET1 EGG                61.0            2.2   
908  BUTTER; SALTED                1 TBSP               14.0            7.1   
909  BUTTER; UNSALTED              1 TBSP               14.0            7.1   
333  POUND CAKE; COMMERCIAL        1 SLICE              29.0            3.0   
913  BUTTER; UNSALTED              1 PAT                 5.0            2.5   
912  BUTTER; SALTED                1 PAT                 5.0            2.5   
384  POUND CAKE; COMMERCIAL        1 LOAF              500.0           52.0   
709  BUTTER; SALTED                1/2 CUP             113.0           57.1   

     cholesterol  saturated_fat_per_gram  cholesterol_per_gram  
119          213                0.094118             12.529412  
58           126                0.020000              6.300000  
45           410                0.029412              4.823529  
167          211                0.041304              4.586957  
186          213                0.032000              4.260000  
184          213                0.032000              4.260000  
185          212                0.030000              4.240000  
189          215                0.036066              3.524590  
908           31                0.507143              2.214286  
909           31                0.507143              2.214286  
333           64                0.103448              2.206897  
913           11                0.500000              2.200000  
912           11                0.500000              2.200000  
384         1100                0.104000              2.200000  
709          247                0.505310              2.185841

<p> <em style="color:#3358FF";> 24. Standardize the field saturated_fat_per_gram. Produce a listing of all the food items that are outliers at the high end of the scale. How many food items are outliers at the low end of the scale?</em></p>

In [102]:
# Use the zscore method from the stats class to calculate a standardized value for saturated_fat_per_gram 
nutrition_subset['saturated_fat_per_gram_z'] = stats.zscore(nutrition_subset['saturated_fat_per_gram'])

# Create a subset of the table with records whose saturated_fat_per_gram_z is greater than 3, and are 
# therefore outliers on the high side
nutrition_subset.query('saturated_fat_per_gram_z > 3' )

food item  weight_in_grams  saturated_fat  \
210  CHOCOLATE; BITTER OT BAKING   1 OZ                28.35            9.0   
448  COCONUT; RAW; SHREDDED        1 CUP               80.00           23.8   
492  COCONUT; DRIED; SWEETND;SHREDD1 CUP               93.00           29.3   
576  COCONUT; RAW; PIECE           1 PIECE             45.00           13.4   
709  BUTTER; SALTED                1/2 CUP            113.00           57.1   
710  BUTTER; UNSALTED              1/2 CUP            113.00           57.1   
890  LARD                          1 CUP              205.00           80.4   
898  FATS; COOKING/VEGETBL SHORTENG1 TBSP              13.00            3.3   
899  LARD                          1 TBSP              13.00            5.1   
907  FATS; COOKING/VEGETBL SHORTENG1 CUP              205.00           51.3   
908  BUTTER; SALTED                1 TBSP              14.00            7.1   
909  BUTTER; UNSALTED              1 TBSP              14.00            7.1   
912  BUTTER; SALTED                1 PAT                5.00            2.5   
913  BUTTER; UNSALTED              1 PAT                5.00            2.5   
920  IMITATION CREAMERS; POWDERED  1 TSP                2.00            0.7   

     cholesterol  saturated_fat_per_gram  cholesterol_per_gram  \
210            0                0.317460              0.000000   
448            0                0.297500              0.000000   
492            0                0.315054              0.000000   
576            0                0.297778              0.000000   
709          247                0.505310              2.185841   
710          247                0.505310              2.185841   
890          195                0.392195              0.951220   
898            0                0.253846              0.000000   
899           12                0.392308              0.923077   
907            0                0.250244              0.000000   
908           31                0.507143              2.214286   
909           31                0.507143              2.214286   
912           11                0.500000              2.200000   
913           11                0.500000              2.200000   
920            0                0.350000              0.000000   

     saturated_fat_per_gram_z  
210                  4.240676  
448                  3.938687  
492                  4.204266  
576                  3.942889  
709                  7.082741  
710                  7.082741  
890                  5.371375  
898                  3.278227  
899                  5.373078  
907                  3.223726  
908                  7.110475  
909                  7.110475  
912                  7.002408  
913                  7.002408  
920                  4.732985

In [103]:
# Create a subset of the data showing the outliers on the low side. Note that this doesn't return any items
nutrition_subset.query('saturated_fat_per_gram_z < -3' )

Empty DataFrame
Columns: [food item, weight_in_grams, saturated_fat, cholesterol, saturated_fat_per_gram, cholesterol_per_gram, saturated_fat_per_gram_z]
Index: []

In [104]:
# Because no records were shown for the low side outliers, we want to see what the lowest value actually is, so we look at the 
# tail of the sorted table. We could have also set ascending = True and accomplished the same thing by reviewing the head.
nutrition_subset_sort = nutrition_subset.sort_values(['saturated_fat_per_gram_z'], ascending = False)
nutrition_subset_sort.tail(n=5)

food item  weight_in_grams  saturated_fat  \
731  TOMATOES; RAW                 1 TOMATO            123.0            0.0   
730  RASPBERRIES; RAW              1 CUP               123.0            0.0   
727  TOMATO JUICE; CANNED W/O SALT 1 CUP               244.0            0.0   
726  TOMATO JUICE; CANNED WITH SALT1 CUP               244.0            0.0   
960  SALT                          1 TSP                 5.5            0.0   

     cholesterol  saturated_fat_per_gram  cholesterol_per_gram  \
731            0                     0.0                   0.0   
730            0                     0.0                   0.0   
727            0                     0.0                   0.0   
726            0                     0.0                   0.0   
960            0                     0.0                   0.0   

     saturated_fat_per_gram_z  
731                 -0.562335  
730                 -0.562335  
727                 -0.562335  
726                 -0.562335  
960                 -0.562335

<p style="color:#2CC531"> We can see there are 15 items that are outliers on the high end of the scale. When performing the query for z-scores less than -3, no results are returned. If we review the tail of the table, we will see that the lowest value for the saturated_fat_per_gram_z z-score is -0.562335. Since this value is above -3, we would conclude there are no outliers on the low end of the data set.</p>

<p><em style="color:#3358FF";>25. Standardize the field cholesterol_per_gram. Produce a listing of all the food items that are outliers at the high end of the scale.</em></p>

In [106]:
# Using the same method I did previously to calculate the cholesterol_per_gram_z value
nutrition_subset['cholesterol_per_gram_z'] = stats.zscore(nutrition_subset['cholesterol_per_gram'])
nutrition_subset_q = nutrition_subset.query('cholesterol_per_gram_z > 3')

# I have created a subset of the table containing only values whose cholesterol_per_gram_z is above 3. I then sort by that value
# and display the sorted table
nutrition_subset_sort = nutrition_subset_q.sort_values(['cholesterol_per_gram_z'], ascending = False)
nutrition_subset_sort.head(n=15)

food item  weight_in_grams  saturated_fat  \
119  EGGS; RAW; YOLK               1 YOLK               17.0            1.6   
58   CHICKEN LIVER; COOKED         1 LIVER              20.0            0.4   
45   BEEF LIVER; FRIED             3 OZ                 85.0            2.5   
167  EGGS; COOKED; FRIED           1 EGG                46.0            1.9   
184  EGGS; RAW; WHOLE              1 EGG                50.0            1.6   
186  EGGS; COOKED; HARD-COOKED     1 EGG                50.0            1.6   
185  EGGS; COOKED; POACHED         1 EGG                50.0            1.5   
189  EGGS; COOKED; SCRAMBLED/OMELET1 EGG                61.0            2.2   

     cholesterol  saturated_fat_per_gram  cholesterol_per_gram  \
119          213                0.094118             12.529412   
58           126                0.020000              6.300000   
45           410                0.029412              4.823529   
167          211                0.041304              4.586957   
184          213                0.032000              4.260000   
186          213                0.032000              4.260000   
185          212                0.030000              4.240000   
189          215                0.036066              3.524590   

     saturated_fat_per_gram_z  cholesterol_per_gram_z  
119                  0.861616               18.179372  
58                  -0.259746                8.952391  
45                  -0.117350                6.765448  
167                  0.062578                6.415037  
184                 -0.078192                5.930750  
186                 -0.078192                5.930750  
185                 -0.108451                5.901127  
189                 -0.016682                4.841464